Support for specifying .gitattributes file in the Init and Clone commands

cmwoods opened the pull request:
While the project already supports specification of a .gitignore file, support for .gitattributes was curiously missing. It is particularly helpful to be able to include a .gitattributes file for the Clone operation in order to address line-ending normalization, LFS, etc...

siprbaum reviewed (changes requested):
You have a quite strange merge history here, either rebase to clean up the history and fix the review findings or we need to do a squash merge *iff* this gets merged.

Anyhow, adding this to the commands makes this support of the .gitattributes kind of official, which I feal a little uneasy about. 
Looking at https://github.com/git-tfs/git-tfs/blob/f60d7e762f6e47ff74bb684b4eabeb39297c2c87/src/GitTfs/Core/GitRepository.cs#L544 we are passing at least one invalid hint path to the underlying method dealing with the .gitattributes and the filter, so I can't tell what the actual state of support for .gitattributes really is.

cmwoods commented on src/CommonAssemblyInfo.cs:
```diff
@@ -1,9 +1,14 @@
-﻿using System.Reflection;
+﻿//------------------------------------------------------------------------------
```
Hmm... Looks like that might of snuck in when I was trying to use cake. I can remove.

cmwoods replied:
Done

cmwoods commented:
> You have a quite strange merge history here, either rebase to clean up the history and fix the review findings or we need to do a squash merge _iff_ this gets merged.

I can see what I can do in this area to clean that up / squash that down on my side. I think it was due to the order of operations in which I was trying to isolate just these changes from other changes that I have in my fork and merge in the latest revisions from git-tfs/master.

> Anyhow, adding this to the commands makes this support of the .gitattributes kind of official, which I feal a little uneasy about.
> Looking at
> 
> https://github.com/git-tfs/git-tfs/blob/f60d7e762f6e47ff74bb684b4eabeb39297c2c87/src/GitTfs/Core/GitRepository.cs#L544
> 
> we are passing at least one invalid hint path to the underlying method dealing with the .gitattributes and the filter, so I can't tell what the actual state of support for .gitattributes really is.

Hmm... Obviously I don't have as deep a working understanding of the codebase. So you're saying that FilteringOptions should be getting the path to the .gitattributes file? If so I can fix that up. If there's more if you can point me in the right direction then I'll see what I can do to address.

siprbaum commented:
AFAIK yes, but I am also not very familiar with the code here. But from what I understand, if you define different attributes depending on the file name, .e.g something like the following in  the `.gitattributes`
```
*.bat	text eol=crlf
*.sh		text eol=lf
``` 
and then having `batchfile.bat` and `shellscript.sh`,  the .gitattributes file tells that the batchfile.bat should be stored in the working directory with CRLF line ending and shellscript.sh file with LF line ending, the class `FilteringOptions` contains a HintPath property (see https://github.com/libgit2/libgit2sharp/blob/df3b22a754ef56da8d7e3c330ce2d783c2b7982e/LibGit2Sharp/FilteringOptions.cs#L25) which is IMHO used to detect which filter should be applied, the one for the `*.bat` or the one for `*.sh`, depending on the filename. To my understanding, this would also be reflected back into TFS as soon as one makes changes locally in git in one of those files. Both would be stored in git with LF line ending, but as soon as one creates a working directory or a workspace for TFS to checkin those files, they would be stored with the configured line ending in the .gitattributes file.

But as I said, I didn't test this and I also didn't check if the code path using the method even cares about using the correct filter, but all in all this make me feel a little nervous when fully introducing .gitattributes.

Pls also have a look at https://github.com/git-tfs/git-tfs/pull/1210 where this was introduced.

siprbaum commented:
So I created a test to confirm my suspicion:
- Create a shelveset with files with differing line endings
- Configure  `core.autocrlf=false` and `.git\info\attributes (repo wide .gitattributes) for a git tfs cloned repo with
  ```
  *.lf eol=lf 
  *.crlf eol=crlf
  ```
- Import the shelveset (to simulate importing from TFS); 
  check that the line endings are all LF in the git object DB when importing as git will do this for all files with a `text` attribute
applied and a `eol` attribute implies `text`. All `*.crlf` files have CRLF and all `*.lf` have LF as line ending in the working directory.

This has tested the import side. 
Now test what happens if these files get back to TFS, e.g. when doing `git tfs rcheckin`. As I don't want to checkin, a sheveset
is created instead, which should show the same behaviour concerning line endings.
- Check the line endings of the files in the prepared TFS workspace (`.git/~w/`) before shelving. As we have .gitattributes all
  `*.lf` files should have LF and all `*.crlf` files should have CRLF line ending.
- Paranoia check (should be unnecessary, but I did it anyhow for other reasons):
  * Re-Import this exported shelveset, but this time without any .gitattributes (.git/info/attributes) file to ensure the file is
     imported as is without any filter/line ending converions applied
  * Check that the files have the expected line endings

If I run the test (via the attached script, saved locally as t.sh) like this  `./t.sh /c/Temp/Service attributes_test /c/tools/GitTfs-0.32.0/git-tfs.exe` then I see an error that the line ending for the `*.crlf` files is actually not CRLF but LF.

if I make the change in https://github.com/git-tfs/git-tfs/blob/f60d7e762f6e47ff74bb684b4eabeb39297c2c87/src/GitTfs/Core/GitRepository.cs#L537
like proposed earlier
```diff csharp
        public void CopyBlob(string sha, string outputFile)
        {
            Blob blob;
            var destination = new FileInfo(outputFile);
            if (!destination.Directory.Exists)
                destination.Directory.Create();
            if ((blob = _repository.Lookup<Blob>(sha)) != null)
-                using (Stream stream = blob.GetContentStream(new FilteringOptions(string.Empty)))
+                using (Stream stream = blob.GetContentStream(new FilteringOptions(outputFile)))
                using (var outstream = File.Create(destination.FullName))
                    stream.CopyTo(outstream);
        }
```
and rerun the script with a git-tfs with that change included, e.g. `./t.sh /c/Temp/Service testme /c/src/git-tfs/src/GitTfs/bin/Debug/net472/git-tfs.exe` then the files are exported correctly with the configured line ending.

Nevertheless, what I couldn't get working is to use a clean & smudge filter ala https://git-scm.com/docs/gitattributes#_filter 
so I think .gitattributes is not fully working.

So I am leaning to accept your changes, but only if it also contains the above fix and a huge disclaimer in our help and usage text that gitattributes are only partially supported and that currently only line endings where tested.

But lets give @pmiossec or @spraints a chance to give feedback, as overall this whole .gitattributes will introduce yet another way to hang yourself when interacting with the TFS.


<details>
<summary>Script for testing line endings and gitattributes</summary>

```bash
#!/bin/sh
# Dummy test script testing gitattributes and lineendings conversion during import and export.
# It uses a prepared shelvesets to import (as a way to simulate importing from TFS) and also tests exporting to a
# shelveset with configured gitattributes and tests what gets synced back to TFS.


trap -- 'rm -f ${clone}/{actual,expected}.txt ${clone}/blob_content_*.txt ${clone}/{,re}imported_test_original_has*'	EXIT

die () {
	echo "ERROR: $@"
	exit 1
}

# delete a git branch if it exists
function delete_git_branch() {
	branch=$1
	git rev-parse --quiet --verify ${branch} && git branch -D ${branch}

}

function show_eol() {
	file=$1
	#LANG=C file $file|grep --quiet 'with CRLF line terminators'
	eol="ERROR: File does not exist"
	if [ -e $file ]; then
		eol=$(dos2unix < $file | cmp --quiet - $file && echo "LF" || echo "CRLF")
	fi

	echo $eol
}

if test $# != 3; then
	die "Usage: <tfs_clone> <shelvsetname> <path_to_git-tfs.exe>"
fi

# Make sure you have a backup copy as it modifies .git/info/attributes and .git/config settings
clone=$1

# name of the shelveset which will be created by the user
shelveset_name=$2

# full path to git-tfs.exe, e.g. /c/tools/GitTfs-0.32.0/git-tfs.exe   or   git_tfs=/c/src/git-tfs/src/GitTfs/bin/Debug/net472/git-tfs.exe
 git_tfs=$3


if [ ! -e ${git_tfs} ]; then
	die "git-tfs not found at ${git_tfs}"
fi

cd ${clone} &&
git config core.autocrlf false 
cat > .git/info/attributes << EOF
*.crlf eol=crlf
*.lf   eol=lf
EOF

# Create a branch where we prepare some files which get shelved
git checkout -B test_filter_prepared tfs/default

# create some files with CRLF and LF in them. The extension represents
# the .gitattribute eol=<extension> which will be later configured when importing.
printf "some\r\ntext\r\n" > test_original_has_crlf.crlf
printf "some\r\ntext\r\n" > test_original_has_crlf.lf
printf "some\ntext\n"     > test_original_has_lf.crlf
printf "some\ntext\n"     > test_original_has_lf.lf

git add test_original_has* &&
git commit -m "Prepared files with differing line endings" &&

${git_tfs} shelve --force ${shelveset_name} test_filter_prepared &&

# Create branch base where we import on top
git checkout -B test_filter tfs/default

# Delete existing imported_shelveset branch from previous run of this script
delete_git_branch "${shelveset_name}_imported"

# Import the shelveset
${git_tfs} unshelve --force ${shelveset_name} ${shelveset_name}_imported &&

# List the files which should be filtered	
git ls-tree ${shelveset_name}_imported|grep test|awk '{print $3 " " $4 }'|while read blob filename
do
	git cat-file -p $blob > imported_${filename}
	printf "%-4s %s\n" $(show_eol imported_${filename}) ${filename}
done > blob_content_actual.txt

cat > blob_content_expected.txt << EOF
LF   test_original_has_crlf.crlf
LF   test_original_has_crlf.lf
LF   test_original_has_lf.crlf
LF   test_original_has_lf.lf
EOF

# For any file whith a 'text' attribute (which is implied by 'eol=lf' or 'eol=crlf')
# the content in the git object database should have LF only.
diff -u blob_content_actual.txt blob_content_expected.txt || die "Line endings differ in blob content after first import"

# Cleanup git-tfs workspace;  shouldn't be needed in practice
rm -rf "${clone}/.git/~w"

############################################
# Export to a shelveset to simulate checkin
############################################
${git_tfs} shelve --force -m 'Shelve/Unshelve imported changes' ${shelveset_name}_exported "${shelveset_name}_imported"

# The following section checks the line ending which is used for exporting it to TFS
for file in .git/~w/test_original_has_crlf.crlf .git/~w/test_original_has_lf.crlf .git/~w/test_original_has_crlf.lf .git/~w/test_original_has_lf.lf; do
	printf "%-4s %s\n" $(show_eol $file) $file
done > actual.txt

cat > expected.txt << EOF
CRLF .git/~w/test_original_has_crlf.crlf
CRLF .git/~w/test_original_has_lf.crlf
LF   .git/~w/test_original_has_crlf.lf
LF   .git/~w/test_original_has_lf.lf
EOF
# Check that the files in .git\~w which are used to create the shelveset if they have the expected line ending
diff -u actual.txt expected.txt || die "Line endings differ in .git/~w when exporting"

# Cleanup git-tfs workspace;  shouldn't be needed in practice
rm -rf "${clone}/.git/~w"

#######################################################################
# Re-Import already exported shelveset, but now without any attributes.
# This tests that we import as-is.
#######################################################################
git checkout -f test_filter &&
rm -f ${clone}/.git/info/attributes &&
delete_git_branch ${shelveset_name}_exported_imported
${git_tfs} unshelve --force ${shelveset_name}_exported ${shelveset_name}_exported_imported

# List the files which should be filtered	
git ls-tree ${shelveset_name}_exported_imported|grep test|awk '{print $3 " " $4 }'|while read blob filename
do
	git cat-file -p $blob > reimported_${filename}
	printf "%-4s %s\n" $(show_eol reimported_${filename}) ${filename}
done > blob_content_reimported_actual.txt

cat > blob_content_reimported_expected.txt << EOF
CRLF test_original_has_crlf.crlf
LF   test_original_has_crlf.lf
CRLF test_original_has_lf.crlf
LF   test_original_has_lf.lf
EOF

diff -u blob_content_reimported_actual.txt blob_content_reimported_expected.txt || die "Line endings differ in blob content after re-importing"

```
</details>

spraints commented on src/GitTfs/Commands/Init.cs:
```diff
@@ -32,7 +33,11 @@ public Init(RemoteOptions remoteOptions, InitOptions initOptions, Globals global
+            get
+            {
+                return _initOptions.OptionSet.Merge(_remoteOptions.OptionSet)
+                    .Add("gitattributes=", "Path to the .gitattributes file to include", v => _gitAttributesFilePath = v);
```
This option is similar to `--gitignore`. Can the definition of the option move into `_remoteOptions`?

siprbaum replied:
Actually I had exactly the same idea. But the RemoteOptions are used in quite a lot of commands where they are also
participate in the commands usage/help text with the code you quoted, e.g. that merging of the options like
```csharp
return _initOptions.OptionSet.Merge(_remoteOptions.OptionSet)
```
So this would mean that all commands would get that additional commandline parameter `--gitattributes=<file>`
which merge the `_remoteOptions.OptionSet`. From my look, the following commands would receive that additional parameter:
* Bootstrap
* Clone (indirectly via Init)
* Fetch / QuickFetch
* Create  (indirectly via Clone)
* Init 
* ListRemoteBranches
* Subtree (indrectly via Fetch)
* Verify

To me, this new parameter makes only sense for Init and Clone actually a repo is created and we `git add` that new file.
Keep in mind that the --gitignore flag sometimes (e.g. in Fetch case) only participates by adding only ignores for the
current run and not permantely by `git add`ing the given gitignore to the repo. This is different for the gitattribues,
because AFAIK there is no underlying mechanism to temporarirly add attributes without merging .git/info/attributes and the .gitattributes file toegether.

To sum it up, I don't think this should be added to the _remoteOptions

spraints commented:
> if I make the change in ... \[the use of`FilteringOptions`] ...

This change looks good to me. It's been a while since I've dealt with CRLF issues, so I might be missing something, but this looks like the right change.



> Nevertheless, what I couldn't get working is to use a clean & smudge filter ala https://git-scm.com/docs/gitattributes#_filter
> so I think .gitattributes is not fully working.

Git-tfs uses libgit2, and libgit2 leaves some things partially implemented. https://github.com/libgit2/libgit2/issues/1473 says "Clean and smudge filtering hooks are supported now in libgit2, although libgit2 will not shell out to external programs on your behalf." I suspect that git-tfs doesn't have any code for shelling out either, which is why it's not working.

siprbaum commented:
@cmwoods Are you willing to take a shoot at this, now that the analysis was done and we figured what is missing?

So for me, this would all boil down to adding the following changes on top of yours
* implement the review comments (no commiting autogenerated code; remove useless usings)
* make the change in src/GitTfs/Core/GitRepository.cs propsed in comment https://github.com/git-tfs/git-tfs/pull/1370#issuecomment-818032860 above
* extending the documentation and the usage string that .gitattributes are only partially implemented, e.g. only line endings and no clean/smudge filters
* getting rid of the useless merges introduced by your changes (this is optional, as we can also squash merge the PR)

siprbaum commented:
In https://github.com/git-tfs/git-tfs/blob/master/src/GitTfsTest/Integration/CloneTests.cs#L309-L338 are already existing tests for line normalization.
A test checking that the .gitattributes are taken correctly can be implemented similarly to those, e.g.
clone with autocrlf=false but with a .gitattribute file like this
```.gitattribute
*.bat     eol=crlf
*.sh       eol=lf
```
and then effectively checking that all the files (*.sh and *.bat) in the index have lf line ending.
In the workspace all *.bat files have crlf and all *.sh files have lf line endings, which should also be checked.
That would effectively test the suggested fix in https://github.com/git-tfs/git-tfs/pull/1370#issuecomment-818032860 with the
`FilteringOptions` getting the correct parameter.

cmwoods commented:
Hmm... I cleaned everything up as requested (I believe), saved off the commits to a new branch, rebased my master back to the original fork point, force pushed that, fetched the latest commits from git-tfs, rebased my branch, merged them back to my master, and pushed.

It seems the force push back to the earlier commit point closed this PR. The commit tab of git-hub says that additional commits will re-open the PR however the push of my changes does not seem to have done that.

cmwoods commented on src/GitTfs/Core/GitRepository.cs:
```diff
@@ -9,6 +9,7 @@
 using GitTfs.Commands;
 using Branch = LibGit2Sharp.Branch;
 using GitTfs.Util;
+using NLog;
```
Done

siprbaum commented:
I don't have the time right now to do a full review, but I would expect at least  some tests of this
new functionallity, like suggested here https://github.com/git-tfs/git-tfs/pull/1370#issuecomment-948399671
so we don't regress in the future.

cmwoods commented:
I've written up two tests - one with --gitignore and one without. Unfortunately (or actually fortunately) the tests failed on 2 out of the 6 different line ending files in my test. One of the files has an incorrect line ending in the index and the other has incorrect line ending in the index and the working tree (after git rm -rf .; git reset --hard HEAD). Checking the git repo generated by the test confirms this and the 2 files come up as containing un-staged changes (which are just the line ending corrections due to the post commit git command line operation correctly applying the .gitattributes file.

It seems that perhaps libgit2sharp (or libgit2) is not 100% recognizing the addition of the .gitattributes file that becomes part of the first commit. Currently, I'm left wondering:
- Is this a subtle bug in how the initial .gitattributes (and/or .gitignore) commit is constructed?
- Could this be a limitation/bug in the particular version of the libgit2sharp library used in the project? Of the libgit2sharp/libgit2 libraries overall?
- Will .gitattributes be correctly handled/applied if it appears in a chain of changesets being cloned from TFS? For example, if 3 changesets into the clone a .gitattributes (and/or .gitignore) file is encountered in a changeet. Are those files correctly handled/applied to the commit itself and all subsequent commits performed during the remainder of the cloning operation?

This last question/concern actually is unrelated to the changes contained in this specific PR to inject a .gitattributes file before the cloning starts but rather may turn out a lurking bug/limitation and/or could provide useful information to solve the problem.

siprbaum commented:
> I've written up two tests - one with --gitignore and one without. Unfortunately (or actually fortunately) the tests failed on 2 out of the 6 different line ending files in my test. One of the files has an incorrect line ending in the index and the other has incorrect line ending in the index and the working tree (after git rm -rf .; git reset --hard HEAD). Checking the git repo generated by the test confirms this and the 2 files come up as containing un-staged changes (which are just the line ending corrections due to the post commit git command line operation correctly applying the .gitattributes file.
> 
> It seems that perhaps libgit2sharp (or libgit2) is not 100% recognizing the addition of the .gitattributes file that becomes part of the first commit. Currently, I'm left wondering:
> 
> * Is this a subtle bug in how the initial .gitattributes (and/or .gitignore) commit is constructed?
> * Could this be a limitation/bug in the particular version of the libgit2sharp library used in the project? Of the libgit2sharp/libgit2 libraries overall?

Not sure on both counts there.  Could you share your tests so I can have a look and get a better understanding
on what is not working?
AFAIK git has problems too when changing the line ending attribute of an already checked in file, so 


> * Will .gitattributes be correctly handled/applied if it appears in a chain of changesets being cloned from TFS? For example, if 3 changesets into the clone a .gitattributes (and/or .gitignore) file is encountered in a changeet. Are those files correctly handled/applied to the commit itself and all subsequent commits performed during the remainder of the cloning operation?
> 

gitattributes support is limited, also in libgit2 but could also be a problem on the git-tfs side.
We should therefore really add the expected behaviour in form of testcases and check if it behaves accordingly and / or within the specification of attributes within git.

> This last question/concern actually is unrelated to the changes contained in this specific PR to inject a .gitattributes file before the cloning starts but rather may turn out a lurking bug/limitation and/or could provide useful information to solve the problem.

cmwoods commented:
I'm closing this PR down unfortunately.

I mistakenly created the PR last year against my fork's master branch which makes it a bit difficult to pull down new commits and/or potentially make other contributions in the future while this work is still pending and since I do not see any way to change the source of the PR to another branch in my fork I've force pushed a reset back to before my changes and resync'd with git-tfs:master.

The work for this PR is still available in [cmwoods:feature/git-attr](https://github.com/cmwoods/git-tfs/tree/feature/git-attr). This branch includes the cleaned up commit history, updates to impacted test's expected hashes, and the requested tests for this functionality - all rebased on top of the latest git-tfs commits as of 2022-03-24. Two of the four new tests still fail.

Should someone figure out the additional changes that are needed, I can always reopen it or create a new PR.
